<center>
<img src="../../img/ods_stickers.jpg" />
    
## [mlcourse.ai](https://mlcourse.ai) – Open Machine Learning Course 
Authors: [Yury Kashnitskiy](https://yorko.github.io) (@yorko), Yury Isakov. Edited by Anna Tarelina (@feuerengel), and Kolchenko Sergey (@KolchenkoSergey). This material is subject to the terms and conditions of the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license. Free use is permitted for any non-commercial purpose.

# <center> Assignment #2. Spring 2019
## <center>  Competition 1. User Identification with Logistic Regression <br>(beating baselines in the "Alice" competition)

    
Today we are going to practice working with sparse matrices, training Logistic Regression models, and doing feature engineering. We will reproduce a couple of baselines in the  Kaggle Inclass competition ["Catch Me If You Can: Intruder Detection through Webpage Session Tracking"](https://www.kaggle.com/c/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking2) (a.k.a. "Alice"). More credits will be given for beating stronger baselines. 

Prior to working on the assignment, you'd better check out the corresponding course material:
 1. [Classification, Decision Trees and k Nearest Neighbors](https://nbviewer.jupyter.org/github/Yorko/mlcourse_open/blob/master/jupyter_english/topic03_decision_trees_kNN/topic3_decision_trees_kNN.ipynb?flush_cache=true), the same as an interactive web-based [Kaggle Kernel](https://www.kaggle.com/kashnitsky/topic-3-decision-trees-and-knn) (basics of machine learning are covered here)
 2. Linear classification and regression in 5 parts: 
    - [ordinary least squares](https://www.kaggle.com/kashnitsky/topic-4-linear-models-part-1-ols)
    - [linear classification](https://www.kaggle.com/kashnitsky/topic-4-linear-models-part-2-classification)
    - [regularization](https://www.kaggle.com/kashnitsky/topic-4-linear-models-part-3-regularization)
    - [logistic regression: pros and cons](https://www.kaggle.com/kashnitsky/topic-4-linear-models-part-4-more-of-logit)
    - [validation](https://www.kaggle.com/kashnitsky/topic-4-linear-models-part-5-validation)
 3. You can also practice with demo assignments, which are simpler and already shared with solutions: 
    - " Sarcasm detection with logistic regression": [assignment](https://www.kaggle.com/kashnitsky/a4-demo-sarcasm-detection-with-logit) + [solution](https://www.kaggle.com/kashnitsky/a4-demo-sarcasm-detection-with-logit-solution)
    - "Linear regression as optimization": [assignment](https://www.kaggle.com/kashnitsky/a4-demo-linear-regression-as-optimization/edit) (solution cannot be officially shared)
    - "Exploring OLS, Lasso and Random Forest in a regression task": [assignment](https://www.kaggle.com/kashnitsky/a6-demo-linear-models-and-rf-for-regression) + [solution](https://www.kaggle.com/kashnitsky/a6-demo-regression-solution)
 4. Alice baseline with logistic regression and "bag of sites", [Kernel](https://www.kaggle.com/kashnitsky/alice-logistic-regression-baseline)
 5. Correct time-aware cross-validation scheme, more features, and hyperparameter optimization, [Kernel](https://www.kaggle.com/kashnitsky/correct-time-aware-cross-validation-scheme)
 6. Other [Kernels](https://www.kaggle.com/c/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking2/kernels?sortBy=voteCount&group=everyone&pageSize=20&competitionId=7173) in this competition. You can share yours as well, but not high-performing ones (Public LB MAE shall be < 0.95). Please don't spoil the competitive spirit.
 7. If that's still not enough, watch two videos on logistic regression: [mlcourse.ai/video](https://mlcourse.ai/video)

**Your task:**
 1. "Follow me". Complete the missing code and submit your answers via [the google form](https://docs.google.com/forms/d/15PVw9CYlX6QnxRHKIDS161kGAq3v7iiO15W3qKTePEY). Use **the same email** as in A1 (for newcomers: remember your email and use it for all forms during the course). 12 credits max. for this part
 2. "Freeride". Come up with good features to beat the baselines "A2 baseline (10 credits)" and "A2 strong baseline (20 credits)". As names suggest, you'll get 10 more credits for beating the first one, and 10 more (20 in total) for beating the second one. You need to name your [team](https://www.kaggle.com/c/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking2/team) (out of 1 person) in full accordance with the [course rating](https://docs.google.com/spreadsheets/d/1LAy1eK8vIONzIWgcCEaVmhKPSj579zK5lrECf_tQT60/edit?usp=sharing) (for newcomers: you need to name your team with your real full name). You can think of it as a part of the assignment.
 3. If you've beaten "A2 baseline (10 credits)" or performed better, you need to upload your solution as described in [course roadmap](https://mlcourse.ai/roadmap) ("Kaggle Inclass Competition Alice" -> Rules). For all baselines that you see on Public Leaderboard, it's OK to beat them on Public LB as well. But 10 winners will be defined according to the private LB, which will be revealed by @yorko on March 11. 
 
 ### <center> Deadline for A2: 2019 March 10, 20:59 GMT (London time)
    
### How to get help
In [ODS Slack](https://opendatascience.slack.com) (if you still don't have access, fill in the [form](https://docs.google.com/forms/d/1BMqcUc-hIQXa0HB_Q2Oa8vWBtGHXk8a6xo5gPnMKYKA/edit) mentioned on the mlcourse.ai main page), we have a channel **#mlcourse_ai_news** with announcements from the course team.
You can discuss the course content freely in the **#mlcourse_ai** channel (we still have a huge Russian-speaking group, they have a separate channel **#mlcourse_ai_rus**).

Please stick to special threads for your questions:
 - [#a2_q1](https://opendatascience.slack.com/archives/C91N8TL83/p1549882527050500) for question 1 below
 - ...
 - [#a2_q8](https://opendatascience.slack.com/archives/C91N8TL83/p1549882547051900) for question 8 below
 - [#a2_alice](https://opendatascience.slack.com/archives/C91N8TL83/p1549882561052200) for general questions on the 1st competition
 
Help each other without sharing correct code and answers. Our TA Aditya @ecdrid is there to help (only in the mentioned threads, do not write to him directly).

# Part 1. Follow me

<img src='../../img/followme_alice.png' width=50%>

*image credit [@muradosmann](https://www.instagram.com/muradosmann/?hl=en)*

In [1]:
# Import libraries and set desired options
import pickle
import numpy as np
import pandas as pd
from scipy.sparse import csr_matrix, hstack
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import roc_auc_score
from sklearn.linear_model import LogisticRegression
from matplotlib import pyplot as plt
import seaborn as sns
sns.set()

##### Problem description

In this competition, we'll analyze the sequence of websites consequently visited by a particular person and try to predict whether this person is Alice or someone else. As a metric we will use [ROC AUC](https://en.wikipedia.org/wiki/Receiver_operating_characteristic).

### 1. Data Downloading and Transformation
Register on [Kaggle](www.kaggle.com), if you have not done it before.
Go to the competition [page](https://inclass.kaggle.com/c/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking2) and download the data.

First, read the training and test sets. Then we'll explore the data in hand and do a couple of simple exercises. 

In [2]:
# Read the training and test data sets, change paths if needed
times = ['time%s' % i for i in range(1, 11)]
train_df = pd.read_csv('/Users/aleksandr/Desktop/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking/train_sessions.csv',
                       index_col='session_id', parse_dates=times)
test_df = pd.read_csv('/Users/aleksandr/Desktop/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking/test_sessions.csv',
                      index_col='session_id', parse_dates=times)

# Sort the data by time
train_df = train_df.sort_values(by='time1')

# Look at the first rows of the training set
train_df.head()

site1               time1  site2               time2  site3  \
session_id                                                                
21669          56 2013-01-12 08:05:57   55.0 2013-01-12 08:05:57    NaN   
54843          56 2013-01-12 08:37:23   55.0 2013-01-12 08:37:23   56.0   
77292         946 2013-01-12 08:50:13  946.0 2013-01-12 08:50:14  951.0   
114021        945 2013-01-12 08:50:17  948.0 2013-01-12 08:50:17  949.0   
146670        947 2013-01-12 08:50:20  950.0 2013-01-12 08:50:20  948.0   

                         time3  site4               time4  site5  \
session_id                                                         
21669                      NaT    NaN                 NaT    NaN   
54843      2013-01-12 09:07:07   55.0 2013-01-12 09:07:09    NaN   
77292      2013-01-12 08:50:15  946.0 2013-01-12 08:50:15  946.0   
114021     2013-01-12 08:50:18  948.0 2013-01-12 08:50:18  945.0   
146670     2013-01-12 08:50:20  947.0 2013-01-12 08:50:21  950.0   

                         time5  ...                 time6  site7  \
session_id                      ...                                
21669                      NaT  ...                   NaT    NaN   
54843                      NaT  ...                   NaT    NaN   
77292      2013-01-12 08:50:16  ...   2013-01-12 08:50:16  948.0   
114021     2013-01-12 08:50:18  ...   2013-01-12 08:50:18  947.0   
146670     2013-01-12 08:50:21  ...   2013-01-12 08:50:21  946.0   

                         time7  site8               time8  site9  \
session_id                                                         
21669                      NaT    NaN                 NaT    NaN   
54843                      NaT    NaN                 NaT    NaN   
77292      2013-01-12 08:50:16  784.0 2013-01-12 08:50:16  949.0   
114021     2013-01-12 08:50:19  945.0 2013-01-12 08:50:19  946.0   
146670     2013-01-12 08:50:21  951.0 2013-01-12 08:50:22  946.0   

                         time9 site10              time10 target  
session_id                                                        
21669                      NaT    NaN                 NaT      0  
54843                      NaT    NaN                 NaT      0  
77292      2013-01-12 08:50:17  946.0 2013-01-12 08:50:17      0  
114021     2013-01-12 08:50:19  946.0 2013-01-12 08:50:20      0  
146670     2013-01-12 08:50:22  947.0 2013-01-12 08:50:22      0  

[5 rows x 21 columns]

The training data set contains the following features:

- **site1** – id of the first visited website in the session
- **time1** – visiting time for the first website in the session
- ...
- **site10** – id of the tenth visited website in the session
- **time10** – visiting time for the tenth website in the session
- **target** – target variable, 1 for Alice's sessions, and 0 for the other users' sessions
    
User sessions are chosen in the way that they are shorter than 30 min. long and contain no more than 10 websites. I.e. a session is considered over either if a user has visited 10 websites or if a session has lasted over 30 minutes.

There are some empty values in the table, it means that some sessions contain less than ten websites. Replace empty values with 0 and change columns types to integer. Also load the websites dictionary and check how it looks like:

In [3]:
# Change site1, ..., site10 columns type to integer and fill NA-values with zeros
sites = ['site%s' % i for i in range(1, 11)]
train_df[sites] = train_df[sites].fillna(0).astype(np.uint16)
test_df[sites] = test_df[sites].fillna(0).astype(np.uint16)

# Load websites dictionary
with open(r"/Users/aleksandr/Desktop/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking/site_dic.pkl", "rb") as input_file:
    site_dict = pickle.load(input_file)

# Create dataframe for the dictionary
sites_dict = pd.DataFrame(list(site_dict.keys()), index=list(site_dict.values()), 
                          columns=['site'])
print(u'Websites total:', sites_dict.shape[0])
sites_dict.head()

Websites total: 48371


site
25075              www.abmecatronique.com
13997                     groups.live.com
42436  majeureliguefootball.wordpress.com
30911           cdt46.media.tourinsoft.eu
8104                  www.hdwallpapers.eu

### 2. Brief Exploratory Data Analysis

Before we start training models, we have to perform Exploratory Data Analysis ([EDA](https://en.wikipedia.org/wiki/Exploratory_data_analysis)). Today, we are going to perform a shorter version, but we will use other techniques as we move forward. Let's check which websites in the training data set are the most visited. As you can see, they are Google services and a bioinformatics website (a website with 'zero'-index is our missed values, just ignore it):

In [4]:
# Top websites in the training data set
top_sites = pd.Series(train_df[sites].values.flatten()
                     ).value_counts().sort_values(ascending=False).head(25)
print(top_sites)
sites_dict.loc[top_sites.drop(0).index]

21     123776
0      122730
23      87619
782     77055
22      58258
29      54094
167     46405
780     43841
778     38194
52      36085
812     35178
80      31391
570     30616
55      27812
39      25275
37      23726
30      23495
786     23026
35      22470
76      20922
33      19683
3       18130
77      17270
616     16890
733     15920
dtype: int64


site
21                   www.google.fr
23                  www.google.com
782                 annotathon.org
22                 apis.google.com
29                www.facebook.com
167                   www.bing.com
780         blast.ncbi.nlm.nih.gov
778           www.ncbi.nlm.nih.gov
52             clients1.google.com
812                mail.google.com
80                   s.youtube.com
570                plus.google.com
55   safebrowsing-cache.google.com
39             accounts.google.com
37                     twitter.com
30            platform.twitter.com
786               www.phylogeny.fr
35        s-static.ak.facebook.com
76                 www.youtube.com
33          static.ak.facebook.com
3                   login.live.com
77                    i1.ytimg.com
616                docs.google.com
733            translate.google.fr

##### 1. What kind of websites does Alice visit the most?
*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a2_q1__*

- videohostings
- social networks
- torrent trackers
- news

In [5]:
a = train_df[train_df['target'] == 1]

In [6]:
sites_dict.loc[top_sites.drop(0).index]

site
21                   www.google.fr
23                  www.google.com
782                 annotathon.org
22                 apis.google.com
29                www.facebook.com
167                   www.bing.com
780         blast.ncbi.nlm.nih.gov
778           www.ncbi.nlm.nih.gov
52             clients1.google.com
812                mail.google.com
80                   s.youtube.com
570                plus.google.com
55   safebrowsing-cache.google.com
39             accounts.google.com
37                     twitter.com
30            platform.twitter.com
786               www.phylogeny.fr
35        s-static.ak.facebook.com
76                 www.youtube.com
33          static.ak.facebook.com
3                   login.live.com
77                    i1.ytimg.com
616                docs.google.com
733            translate.google.fr

In [7]:
top_a=pd.Series(a[sites].values.flatten()).value_counts()

In [8]:
a.head(1)

site1               time1  site2               time2  site3  \
session_id                                                                
251175        270 2013-02-12 16:25:10    270 2013-02-12 16:25:11    270   

                         time3  site4               time4  site5  \
session_id                                                         
251175     2013-02-12 16:32:10     21 2013-02-12 16:32:11     21   

                         time5  ...                 time6 site7  \
session_id                      ...                               
251175     2013-02-12 16:32:24  ...   2013-02-12 16:32:25    21   

                         time7 site8               time8 site9  \
session_id                                                       
251175     2013-02-12 16:32:25  7832 2013-02-12 16:32:26    30   

                         time9 site10              time10 target  
session_id                                                        
251175     2013-02-12 16:32:27   7832 2013-02-12 16:32:27      1  

[1 rows x 21 columns]

In [9]:
top_a

77       1382
80       1354
76       1307
29        897
21        857
81        609
879       522
22        522
75        451
82        447
23        437
35        381
881       371
37        293
33        291
3000      286
733       274
30        272
78        236
941       215
7832      209
52        206
704       204
0         201
2078      188
617       159
1057      155
942       153
270       151
335       150
         ... 
24480       1
27294       1
21584       1
8931        1
27370       1
27240       1
27182       1
27386       1
970         1
13934       1
27272       1
188         1
3003        1
12613       1
2590        1
27288       1
2381        1
6400        1
508         1
540         1
27169       1
19190       1
14244       1
27185       1
5061        1
698         1
27352       1
1570        1
5648        1
3159        1
Length: 1054, dtype: int64

In [10]:
sites_dict.loc[top_a.drop(0).index]

site
77                                          i1.ytimg.com
80                                         s.youtube.com
76                                       www.youtube.com
29                                      www.facebook.com
21                                         www.google.fr
81                  r4---sn-gxo5uxg-jqbe.googlevideo.com
879                 r1---sn-gxo5uxg-jqbe.googlevideo.com
22                                       apis.google.com
75                                           s.ytimg.com
82                  r2---sn-gxo5uxg-jqbe.googlevideo.com
23                                        www.google.com
35                              s-static.ak.facebook.com
881                 r3---sn-gxo5uxg-jqbe.googlevideo.com
37                                           twitter.com
33                                static.ak.facebook.com
3000                                              vk.com
733                                  translate.google.fr
30                                  platform.twitter.com
78                                         yt3.ggpht.com
941                                      mts0.google.com
7832                                 www.info-jeunes.net
52                                   clients1.google.com
704                             www.audienceinsights.net
2078                                        www.melty.fr
617                                        gg.google.com
1057                                 plus.googleapis.com
942                                      mts1.google.com
270                                         api.bing.com
335                                  www.dailymotion.com
12619                                       youwatch.org
...                                                  ...
24480                                 images.hellopro.fr
27294  i1-js-14-3-01-10077-975669919-i.init.cedexis-r...
21584                               www.forma-search.com
8931                                    dpm.zebestof.com
27370                                www.justice.gouv.fr
27240                             jeuxstrategie1.free.fr
27182                                api.kickstarter.com
27386                           banners.netopartners.com
970                             dtm.thingsremembered.com
13934                   event-tracking.api.jobrapido.com
27272  i1-js-14-3-01-10165-192443829-i.init.cedexis-r...
188                                       www.hotmail.fr
3003                                chart.googleapis.com
12613                                   main.vodonet.net
2590                                    www.20minutes.fr
27288                                    www.furplay.net
2381                       d297h9he240fqh.cloudfront.net
6400                                     ct.buzzfeed.com
508                                  assets1.lefigaro.fr
540                              casale-match.dotomi.com
27169                                  fs30.youwatch.org
19190                                   ptracker.be2.com
14244                                  www.gamesquat.com
27185                                        mg.dt07.net
5061                          tracking.shopping-flux.com
698                             a406.idata.over-blog.com
27352                                       www4.cbox.ws
1570                                    static.123rf.net
5648                                     a3.mzstatic.com
3159                               geoip.radio-canada.ca

[1053 rows x 1 columns]

In [11]:
# VIDEOHOSTINGS

Now let us look at the timestamps and try to characterize sessions as timeframes:

In [12]:
# Create a separate dataframe where we will work with timestamps
time_df = pd.DataFrame(index=train_df.index)
time_df['target'] = train_df['target']

# Find sessions' starting and ending
time_df['min'] = train_df[times].min(axis=1)
time_df['max'] = train_df[times].max(axis=1)

# Calculate sessions' duration in seconds
time_df['seconds'] = (time_df['max'] - time_df['min']) / np.timedelta64(1, 's')

time_df.head()

target                 min                 max  seconds
session_id                                                         
21669            0 2013-01-12 08:05:57 2013-01-12 08:05:57      0.0
54843            0 2013-01-12 08:37:23 2013-01-12 09:07:09   1786.0
77292            0 2013-01-12 08:50:13 2013-01-12 08:50:17      4.0
114021           0 2013-01-12 08:50:17 2013-01-12 08:50:20      3.0
146670           0 2013-01-12 08:50:20 2013-01-12 08:50:22      2.0

In [13]:
(time_df['target'] ==1).value_counts()

False    251264
True       2297
Name: target, dtype: int64

In [14]:
alice_time = time_df[time_df['target'] ==1]

In [15]:
other_time = time_df[time_df['target'] ==0]

In [16]:
alice_time['seconds'].mean()

52.29647366129734

In [17]:
other_time['seconds'].mean()

139.28237232552215

In [18]:
time_df.target.value_counts()

0    251264
1      2297
Name: target, dtype: int64

In [19]:
2297/251264 * 100

0.9141779164544065

In [20]:
alice_time['seconds'].min()

0.0

In [21]:
alice_time['seconds'].max()

1763.0

In [22]:
other_time['seconds'].min()

0.0

In [23]:
other_time['seconds'].max()

1800.0

In [24]:
alice_time['seconds'].std()

153.30901405274335

In [25]:
other_time['seconds'].std()

296.65351771850226

In [26]:
(alice_time['seconds'] <41).value_counts()

True     1758
False     539
Name: seconds, dtype: int64

In [27]:
539/(539+1758)

0.2346538963865912

In order to perform the next task, generate descriptive statistics as you did in the first assignment.

*In the next question, we are using the notion of "approximately the same". To be strict, let's define it: $a$ is approximately the same as $b$ ($a \approx b $) if their difference is less than or equal to 5% of the maximum between $a$ and $b$, i.e. $a \approx b \leftrightarrow \frac{|a-b|}{max(a,b)} \leq 0.05$.*

##### 2. Select all correct statements:
*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a2_q2__*

- on average, Alice's session is shorter than that of other users
- more than 1% of all sessions in the dataset belong to Alice
- minimum and maximum durations of Alice's and other users' sessions are approximately the same
- standard deviation of Alice's sessions duration is approximately the same as for non-Alice's sessions
- less than a quarter of Alice's sessions are greater than or equal to 40 seconds

In [28]:
# 1 da 2 net 3 da? 4 net 5 da 

In order to train our first model, we need to prepare the data. First of all, exclude the target variable from the training set. Now both training and test sets have the same number of columns, therefore aggregate them into one dataframe.  Thus, all transformations will be performed simultaneously on both training and test data sets. 

On the one hand, it leads to the fact that both data sets have one feature space (you don't have to worry that you forgot to transform a feature in some data sets). On the other hand, processing time will increase. 
For the enormously large sets it might turn out that it is impossible to transform both data sets simultaneously (and sometimes you have to split your transformations into several stages only for train/test data set).
In our case, with this particular data set, we are going to perform all the transformations for the whole united dataframe at once, and before training the model or making predictions we will just take its appropriate part.

In [29]:
# Our target variable
y_train = train_df['target']

# United dataframe of the initial data 
full_df = pd.concat([train_df.drop('target', axis=1), test_df])

# Index to split the training and test data sets
idx_split = train_df.shape[0]

For the very basic model, we will use only the visited websites in the session (but we will not take into account timestamp features). The point behind this data selection is: *Alice has her favorite sites, and the more often you see these sites in the session, the higher probability that this is Alice's session, and vice versa.*

Let us prepare the data, we will take only features `site1, site2, ... , site10` from the whole dataframe. Keep in mind that the missing values are replaced with zero. Here is how the first rows of the dataframe look like:

In [30]:
# Dataframe with indices of visited websites in session
full_sites = full_df[sites]
full_sites.head()

site1  site2  site3  site4  site5  site6  site7  site8  site9  \
session_id                                                                  
21669          56     55      0      0      0      0      0      0      0   
54843          56     55     56     55      0      0      0      0      0   
77292         946    946    951    946    946    945    948    784    949   
114021        945    948    949    948    945    946    947    945    946   
146670        947    950    948    947    950    952    946    951    946   

            site10  
session_id          
21669            0  
54843            0  
77292          946  
114021         946  
146670         947

Sessions are sequences of website indices, and data in this representation is useless for machine learning method (just think, what happens if we switched all ids of all websites). 

According to our hypothesis (Alice has favorite websites), we need to transform this dataframe so each website has a corresponding feature (column) and its value is equal to number of this website visits in the session. It can be done in two lines:

In [31]:
# sequence of indices
sites_flatten = full_sites.values.flatten()

# and the matrix we are looking for 
# (make sure you understand which of the `csr_matrix` constructors is used here)
# a further toy example will help you with it
full_sites_sparse = csr_matrix(([1] * sites_flatten.shape[0],
                                sites_flatten,
                                range(0, sites_flatten.shape[0]  + 10, 10)))[:, 1:]

In [32]:
full_sites_sparse.shape

(336358, 48371)

In [33]:
336358*48371/1e9

16.269972818

If you understand what just happened here, then you can skip the next passage (perhaps, you can handle logistic regression too?), If not, then let us figure it out.

### Important detour #1: Sparse Matrices

Let us estimate how much memory it will require to store our data in the example above. Our united dataframe contains 336 thousand samples of 48 thousand integer features in each. It's easy to calculate the required amount of memory, roughly:

$$336\ K * 48\ K * 8\ bytes \approx 16* 10^9 * 8\ bytes = 128\ GB,$$

(that's the [exact](http://www.wolframalpha.com/input/?i=336358*48371*8+bytes) value). Obviously, ordinary mortals have no such volumes (strictly speaking, Python may allow you to create such a matrix, but it will not be easy to do anything with it). The interesting fact is that most of the elements of our matrix are zeros. If we count non-zero elements, then it will be about 1.8 million, i.е. slightly more than 10% of all matrix elements. Such a matrix, where most elements are zeros, is called sparse, and the ratio between the number of zero elements and the total number of elements is called the sparseness of the matrix.

For the work with such matrices you can use `scipy.sparse` library, check [documentation](https://docs.scipy.org/doc/scipy-0.18.1/reference/sparse.html) to understand what possible types of sparse matrices are, how to work with them and in which cases their usage is most effective. You can learn how they are arranged, for example, in Wikipedia [article](https://en.wikipedia.org/wiki/Sparse_matrix).
Note, that a sparse matrix contains only non-zero elements, and you can get the allocated memory size like this (significant memory savings are obvious):

In [34]:
# How much memory does a sparse matrix occupy?
print('{0} elements * {1} bytes = {2} bytes'.format(full_sites_sparse.count_nonzero(), 8, 
                                                    full_sites_sparse.count_nonzero() * 8))
# Or just like this:
print('sparse_matrix_size = {0} bytes'.format(full_sites_sparse.data.nbytes))

1866898 elements * 8 bytes = 14935184 bytes
sparse_matrix_size = 14935184 bytes


Let us explore how the matrix with the websites has been formed using a mini example. Suppose we have the following table with user sessions:

| id | site1 | site2 | site3 |
|---|---|---|---|
| 1 | 1 | 0 | 0 |
| 2 | 1 | 3 | 1 |
| 3 | 2 | 3 | 4 |

There are 3 sessions, and no more than 3 websites in each. Users visited four different sites in total (there are numbers from 1 to 4 in the table cells). And let us assume that the mapping is:

 1. vk.com
 2. habrahabr.ru 
 3. yandex.ru
 4. ods.ai

If the user has visited less than 3 websites during the session, the last few values will be zero. We want to convert the original dataframe in a way that each session has a corresponding row which shows the number of visits to each particular site. I.e. we want to transform the previous table into the following form:

| id | vk.com | habrahabr.ru | yandex.ru | ods.ai |
|---|---|---|---|---|
| 1 | 1 | 0 | 0 | 0 |
| 2 | 2 | 0 | 1 | 0 |
| 3 | 0 | 1 | 1 | 1 |


To do this, use the constructor: `csr_matrix ((data, indices, indptr))` and create a frequency table (see examples, code and comments on the links above to see how it works). Here we set all the parameters explicitly for greater clarity:

In [35]:
# data, create the list of ones, length of which equal to the number of elements in the initial dataframe (9)
# By summing the number of ones in the cell, we get the frequency,
# number of visits to a particular site per session
data = [1] * 9

# To do this, you need to correctly distribute the ones in cells
# Indices - website ids, i.e. columns of a new matrix. We will sum ones up grouping them by sessions (ids)
indices = [1, 0, 0, 1, 3, 1, 2, 3, 4]

# Indices for the division into rows (sessions)
# For example, line 0 is the elements between the indices [0; 3) - the rightmost value is not included
# Line 1 is the elements between the indices [3; 6)
# Line 2 is the elements between the indices [6; 9) 
indptr = [0, 3, 6, 9]

# Aggregate these three variables into a tuple and compose a matrix
# To display this matrix on the screen transform it into the usual "dense" matrix
csr_matrix((data, indices, indptr)).todense()

matrix([[2, 1, 0, 0, 0],
        [0, 2, 0, 1, 0],
        [0, 0, 1, 1, 1]])

As you might have noticed, there are not four columns in the resulting matrix (corresponding to number of different websites) but five. A zero column has been added, which indicates if the session was shorter (in our mini example we took sessions of three). This column is excessive and should be removed from the dataframe (do that yourself).

##### 3. What is the sparseness of the matrix in our small example?
*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a2_q3__*

- 42%
- 47%
- 50%
- 53%

In [36]:
m_sp = csr_matrix((data, indices, indptr))[:,1:]
print('{0}%'.format(m_sp.count_nonzero() / (m_sp.shape[0] * m_sp.shape[1]) * 100))

50.0%


Another benefit of using sparse matrices is that there are special implementations of both matrix operations and machine learning algorithms for them, which sometimes allows to significantly accelerate operations due to the data structure peculiarities. This applies to logistic regression as well. Now everything is ready to build our first model.

### 3. Training the first model

So, we have an algorithm and data for it. Let us build our first model, using [logistic regression](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html) implementation from ` Sklearn` with default parameters. We will use the first 90% of the data for training (the training data set is sorted by time), and the remaining 10% for validation. Let's write a simple function that returns the quality of the model and then train our first classifier:

In [37]:
def get_auc_lr_valid(X, y, C=1.0, seed=17, ratio = 0.9):
    # Split the data into the training and validation sets
    idx = int(round(X.shape[0] * ratio))
    # Classifier training
    lr = LogisticRegression(C=C, random_state=seed, solver='liblinear').fit(X[:idx, :], y[:idx])
    # Prediction for validation set
    y_pred = lr.predict_proba(X[idx:, :])[:, 1]
    # Calculate the quality
    score = roc_auc_score(y[idx:], y_pred)
    
    return score

In [38]:
%%time
# Select the training set from the united dataframe (where we have the answers)
X_train = full_sites_sparse[:idx_split, :]

# Calculate metric on the validation set
print(get_auc_lr_valid(X_train, y_train))

0.9195242567955982
CPU times: user 4.17 s, sys: 97.6 ms, total: 4.27 s
Wall time: 4.4 s


The first model demonstrated the quality  of 0.92 on the validation set. Let's take it as the first baseline and starting point. To make a prediction on the test data set **we need to train the model again on the entire training data set** (until this moment, our model used only part of the data for training), which will increase its generalizing ability:

In [39]:
# Function for writing predictions to a file
def write_to_submission_file(predicted_labels, out_file,
                             target='target', index_label="session_id"):
    predicted_df = pd.DataFrame(predicted_labels,
                                index = np.arange(1, predicted_labels.shape[0] + 1),
                                columns=[target])
    predicted_df.to_csv(out_file, index_label=index_label)

In [40]:
# Train the model on the whole training data set
# Use random_state=17 for repeatability
# Parameter C=1 by default, but here we set it explicitly
lr = LogisticRegression(C=1.0, random_state=17, solver='liblinear').fit(X_train, y_train)

# Make a prediction for test data set
X_test = full_sites_sparse[idx_split:,:]
y_test = lr.predict_proba(X_test)[:, 1]

# Write it to the file which could be submitted
write_to_submission_file(y_test, 'baseline_1.csv')

If you follow these steps and upload the answer to the competition [page](https://inclass.kaggle.com/c/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking2), you will get `ROC AUC = 0.90812` on the public leaderboard ("A2 baseline 1").

### 4. Model Improvement: Feature Engineering

Now we are going to try to improve the quality of our model by adding new features to the data. But first, answer the following question:

##### 4. What years are present in the training and test datasets, if united?
*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a2_q4__*

- 13 and 14
- 2012 and 2013
- 2013 and 2014
- 2014 and 2015

In [41]:
#2013 and 2014?

Create a feature that will be a number in YYYYMM format from the date when the session was held, for example 201407 -- year 2014 and 7th month. Thus, we will take into account the monthly [linear trend](http://people.duke.edu/~rnau/411trend.htm) for the entire period of the data provided.

In [42]:
full_df.index

Int64Index([ 21669,  54843,  77292, 114021, 146670, 242171,  57157, 240201,
            210686,  98804,
            ...
             82788,  82789,  82790,  82791,  82792,  82793,  82794,  82795,
             82796,  82797],
           dtype='int64', name='session_id', length=336358)

In [43]:
alice_time['min']

session_id
251175   2013-02-12 16:25:10
196388   2013-02-12 16:32:27
172448   2013-02-12 16:32:53
70129    2013-02-12 16:33:50
206254   2013-02-12 16:33:55
167235   2013-02-12 16:33:55
54979    2013-02-12 16:34:25
77902    2013-02-12 16:34:40
203387   2013-02-12 16:34:52
104441   2013-02-12 16:35:27
76310    2013-02-12 16:35:50
195629   2013-02-12 16:36:54
213093   2013-02-12 16:37:03
157429   2013-02-12 16:37:19
65043    2013-02-12 16:37:53
78974    2013-02-12 16:38:01
226267   2013-02-12 16:38:40
220102   2013-02-12 16:39:18
138364   2013-02-12 16:39:19
144809   2013-02-12 16:39:33
131623   2013-02-12 16:40:01
223045   2013-02-12 16:40:19
96047    2013-02-12 16:40:29
153722   2013-02-12 16:40:36
141517   2013-02-12 16:43:20
78455    2013-02-12 16:43:31
28854    2013-02-12 16:44:19
137033   2013-02-12 16:44:26
19962    2013-02-12 16:44:38
16185    2013-02-12 16:45:23
                 ...        
70568    2014-04-15 17:29:33
61740    2014-04-15 17:29:44
186166   2014-04-15 17:30:01
265

In [44]:
# Dataframe for new features
full_new_feat = pd.DataFrame(index=full_df.index)

# Add start_month feature
full_new_feat['start_month'] = full_df['time1'].apply(lambda ts: 
                                                      100 * ts.year + ts.month).astype('float64')

In [45]:
alice_time['test'] = alice_time['min'].apply(lambda ts:  100 * ts.year + ts.month + ts.day).astype('float64')

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [46]:
full_new_feat.head()

start_month
session_id             
21669          201301.0
54843          201301.0
77292          201301.0
114021         201301.0
146670         201301.0

In [47]:
alice_time['min'].value_counts()

2014-02-18 12:10:42    2
2013-11-15 13:36:41    2
2013-02-12 16:33:55    2
2014-02-18 12:08:55    2
2014-02-17 18:07:50    2
2014-02-18 12:10:44    2
2014-01-28 14:14:19    1
2014-03-25 12:04:22    1
2013-04-12 09:26:15    1
2014-03-24 16:37:20    1
2014-03-24 16:31:10    1
2014-04-15 17:30:19    1
2014-02-17 16:30:07    1
2014-03-25 12:03:26    1
2013-02-12 16:49:30    1
2014-03-24 18:08:10    1
2014-03-24 16:12:10    1
2013-12-16 17:31:15    1
2014-03-24 17:31:18    1
2013-02-12 16:48:50    1
2013-09-12 15:55:50    1
2014-04-14 17:07:51    1
2014-03-24 16:18:54    1
2014-02-17 16:30:14    1
2014-02-20 15:00:28    1
2014-01-04 17:50:03    1
2014-01-04 17:49:55    1
2013-12-17 12:27:53    1
2014-01-20 16:12:12    1
2014-03-24 16:38:30    1
                      ..
2014-02-17 18:04:15    1
2013-09-12 16:45:04    1
2014-02-17 18:03:51    1
2014-03-24 16:07:57    1
2014-02-17 16:36:56    1
2014-03-31 16:48:23    1
2013-09-12 16:44:00    1
2013-12-16 17:36:25    1
2014-03-24 16:06:37    1


In [48]:
alice_time['test'].value_counts()

201321.0    377
201427.0    349
201419.0    218
201418.0    170
201340.0    143
201328.0    106
201330.0    101
201333.0    100
201428.0     97
201426.0     94
201415.0     88
201326.0     65
201329.0     65
201314.0     61
201421.0     59
201405.0     42
201429.0     40
201406.0     38
201316.0     38
201434.0     22
201420.0     13
201417.0      6
201423.0      4
201422.0      1
Name: test, dtype: int64

##### 5. Plot the graph of the number of Alice sessions versus the new feature, start_month. Choose the correct statement:

*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a2_q5__*

- Alice wasn't online at all for the entire period
- From the beginning of 2013 to mid-2014, the number of Alice's sessions per month decreased
- The number of Alice's sessions per month is generally constant for the entire period
- From the beginning of 2013 to mid-2014, the number of Alice's sessions per month increased

*Hint: the graph will be more explicit if you treat `start_month` as a categorical ordinal variable*.

In [49]:
#1 net 2 net 3 net 4 da

In this way, we have an illustration and thoughts about the usefulness of the new feature, add it to the training sample and check the quality of the new model:

In [50]:
# Add the new feature to the sparse matrix
tmp = full_new_feat[['start_month']].values
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], tmp[:idx_split,:]]))

# Compute the metric on the validation set
print(get_auc_lr_valid(X_train, y_train))

0.7508354860175162


In [51]:
full_new_feat

start_month
session_id             
21669          201301.0
54843          201301.0
77292          201301.0
114021         201301.0
146670         201301.0
242171         201301.0
57157          201301.0
240201         201301.0
210686         201301.0
98804          201301.0
113494         201301.0
223837         201301.0
145475         201301.0
186633         201301.0
45193          201301.0
102930         201301.0
15789          201301.0
205868         201301.0
21102          201301.0
193454         201301.0
252924         201301.0
12713          201301.0
182016         201301.0
55750          201301.0
237109         201301.0
120126         201301.0
140105         201301.0
129357         201301.0
136699         201301.0
113189         201301.0
...                 ...
82768          201405.0
82769          201410.0
82770          201405.0
82771          201411.0
82772          201406.0
82773          201406.0
82774          201410.0
82775          201407.0
82776          201405.0
82777          201405.0
82778          201406.0
82779          201405.0
82780          201407.0
82781          201407.0
82782          201408.0
82783          201405.0
82784          201405.0
82785          201406.0
82786          201407.0
82787          201411.0
82788          201412.0
82789          201408.0
82790          201410.0
82791          201405.0
82792          201405.0
82793          201410.0
82794          201405.0
82795          201405.0
82796          201405.0
82797          201411.0

[336358 rows x 1 columns]

The quality of the model has decreased significantly. We added a feature that definitely seemed useful to us, but its usage only worsened the model. Why did it happen?

### Important detour #2: is it necessary to scale features?

Here we give an intuitive reasoning (a rigorous mathematical justification for one or another aspect in linear models you can easily find on the internet). Consider the features more closely: those of them that correspond to the number of visits to a particular web-site per session vary from 0 to 10. The feature `start_month` has a completely different range: from 201301 to 201412, this means the contribution of this variable is significantly greater than the others. It would seem that problem can be avoided if we put less weight in a linear combination of attributes in this case, but in our case logistic regression with regularization is used (by default, this parameter is `C = 1`), which penalizes the model the stronger the greater its weights are. Therefore, for linear methods with regularization, it is recommended to convert features to the same scale (you can read more about the regularization, for example, [here](https://habrahabr.ru/company/ods/blog/322076/)).

One way to do this is standardization: for each observation you need to subtract the average value of the feature and divide this difference by the standard deviation:

$$ x^{*}_{i} = \dfrac{x_{i} - \mu_x}{\sigma_x}$$

The following practical tips can be given:
- It is recommended to scale features if they have essentially different ranges or different units of measurement (for example, the country's population is indicated in units, and the country's GNP in trillions)
- Scale features if you do not have a reason/expert opinion to give a greater weight to any of them
- Scaling can be excessive if the ranges of some of your features differ from each other, but they are in the same system of units (for example, the proportion of middle-aged people and people over 80 among the entire population)
- If you want to get an interpreted model, then build a model without regularization and scaling (most likely, its quality will be worse)
- Binary features (which take only values of 0 or 1) are usually left without conversion, (but)
- If the quality of the model is crucial, try different options and select one where the quality is better

Getting back to `start_month`, let us rescale the new feature and train the model again. This time the quality has increased:

In [52]:
# Add the new standardized feature to the sparse matrix
tmp = StandardScaler().fit_transform(full_new_feat[['start_month']])
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], tmp[:idx_split,:]]))

# Compute metric on the validation set
print(get_auc_lr_valid(X_train, y_train))

0.9196986151568289


In [53]:
# Dataframe for new features
full_new_feat = pd.DataFrame(index=full_df.index)

# Add start_month feature
full_new_feat['start_month'] = full_df['time1'].apply(lambda ts: 
                                                      100 * ts.year + ts.month).astype('float64')

##### 6. Add to the training set a new feature "n_unique_sites" – the number of the unique web-sites in a session. Calculate how the quality on the validation set has changed

*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a2_q6__*

- It has decreased. It is better not to add a new feature.
- It has not changed.
- It has decreased. The new feature should be scaled.
- I am confused, and I do not know if it's necessary to scale a new feature.

*Tips: use the nunique() function from `pandas`. Do not forget to include the start_month in the set. Will you scale a new feature? Why?*

In [ ]:
full_new_feat['n_unique_sites'] = full_df[sites].apply(lambda x: 
                                                       x.nunique(), axis=1).astype('float64')

In [ ]:
full_new_feat

In [ ]:
tmp_raw = full_new_feat[['n_unique_sites']].values

In [ ]:
tmp_scaled = StandardScaler().fit_transform(full_new_feat[['start_month']])

In [ ]:
tmp_scaled

In [ ]:
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], 
                             tmp_scaled[:idx_split,:],
                             tmp_raw[:idx_split,:]]))

In [ ]:
print(get_auc_lr_valid(X_train, y_train))

In [ ]:
tmp_scaled = StandardScaler().fit_transform(full_new_feat[['start_month', 
                                                           'n_unique_sites']])
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], 
                             tmp_scaled[:idx_split,:]]))

In [ ]:
print(get_auc_lr_valid(X_train, y_train))

So, the new feature has slightly decreased the quality, so we will not use it. Nevertheless, do not rush to throw features out because they haven't performed well. They can be useful in a combination with other features (for example, when a new feature is a ratio or a product of two others).

#####  7. Add two new features: start_hour and morning. Calculate the metric. Which of these features gives an improvement?

The `start_hour` feature is the hour at which the session started (from 0 to 23), and the binary feature `morning` is equal to 1 if the session started in the morning and 0 if the session started later (we assume that morning means `start_hour` is equal to 11 or less).

Will you scale the new features? Make your assumptions and test them in practice.

*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a2_q7__*

- None of the features gave an improvement :(
- `start_hour` feature gave an improvement, and `morning` did not
- `morning` feature gave an improvement, and `start_hour` did not
- Both features gave an improvement

*Tip: find suitable functions for working with time series data in [documentation](http://pandas.pydata.org/pandas-docs/stable/api.html). Do not forget to include the `start_month` feature.*

In [ ]:
full_new_feat['start_hour'] = full_df['time1'].apply(lambda ts: ts.hour).astype('float64')
full_new_feat['morning'] = full_new_feat['start_hour'].apply(lambda hour: 
                                                             float(hour <= 11))

In [ ]:
tmp_scaled = StandardScaler().fit_transform(full_new_feat[['start_month', 'start_hour']])
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], 
                             tmp_scaled[:idx_split,:]]))

In [ ]:
print(get_auc_lr_valid(X_train, y_train))

In [ ]:
tmp_raw = full_new_feat[['start_hour']].values
tmp_scaled = StandardScaler().fit_transform(full_new_feat[['start_month']])

In [ ]:
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], 
                             tmp_scaled[:idx_split,:],
                             tmp_raw[:idx_split,:]]))

In [ ]:
print(get_auc_lr_valid(X_train, y_train))

In [ ]:
tmp_scaled = StandardScaler().fit_transform(full_new_feat[['start_month', 'morning']])
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], 
                             tmp_scaled[:idx_split,:]]))

In [ ]:
print(get_auc_lr_valid(X_train, y_train))

In [ ]:
tmp_raw = full_new_feat[['morning']].values
tmp_scaled = StandardScaler().fit_transform(full_new_feat[['start_month']])
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], 
                             tmp_scaled[:idx_split,:],
                             tmp_raw[:idx_split,:]]))

In [ ]:
print(get_auc_lr_valid(X_train, y_train))

In [ ]:
tmp_scaled = StandardScaler().fit_transform(full_new_feat[['start_month', 'start_hour', 
                                                           'morning']])
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], 
                             tmp_scaled[:idx_split,:]]))

In [ ]:
print(get_auc_lr_valid(X_train, y_train))

In [ ]:
# Both features gave an improvement

### 5. Regularization and Parameter Tuning

We have introduced features that improve the quality of our model in comparison with the first baseline. Can we do even better? After we have changed the training and test sets, it almost always makes sense to search for the optimal hyperparameters - the parameters of the model that do not change during training.

For example, in week 3, you learned that, in decision trees, the depth of the tree is a hyperparameter, but the feature by which splitting occurs and its threshold is not. 

In the logistic regression that we use, the weights of each feature are changing, and we find their optimal values during training; meanwhile, the regularization parameter remains constant. This is the hyperparameter that we are going to optimize now.

Calculate the quality on a validation set with a regularization parameter, which is equal to 1 by default:

In [ ]:
# Compose the training set
tmp_scaled = StandardScaler().fit_transform(full_new_feat[['start_month', 
                                                           'start_hour', 
                                                           'morning']])
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], 
                             tmp_scaled[:idx_split,:]]))

# Capture the quality with default parameters
score_C_1 = get_auc_lr_valid(X_train, y_train)
print(score_C_1)

We will try to beat this result by optimizing the regularization parameter. We will take a list of possible values of C and calculate the quality metric on the validation set for each of C-values:

In [ ]:
from tqdm import tqdm

# List of possible C-values
Cs = np.logspace(-3, 1, 10)
scores = []
for C in tqdm(Cs):
    scores.append(get_auc_lr_valid(X_train, y_train, C=C))

Plot the graph of the quality metric (AUC-ROC) versus the value of the regularization parameter. The value of quality metric corresponding to the default value of C=1 is represented by a horizontal dotted line:

In [ ]:
plt.plot(Cs, scores, 'ro-')
plt.xscale('log')
plt.xlabel('C')
plt.ylabel('AUC-ROC')
plt.title('Regularization Parameter Tuning')
# horizontal line -- model quality with default C value
plt.axhline(y=score_C_1, linewidth=.5, color='b', linestyle='dashed') 
plt.show()

##### 8. What is the value of parameter C (if rounded to 2 decimals) that corresponds to the highest model quality?

*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a2_q8__*

- 0.17
- 0.46
- 1.29
- 3.14

In [ ]:
C = Cs[scores.index(max(scores))]
print(C)

In [ ]:
# 0.17

For the last task in this assignment: train the model using the optimal regularization parameter you found (do not round up to two digits like in the last question). If you do everything correctly and submit your solution, you should see `ROC AUC = 0.92784` on the public leaderboard ("A2 baseline 2"):

In [ ]:
full_new_feat


In [ ]:
# Prepare the training and test data
tmp_scaled = StandardScaler().fit_transform(full_new_feat[['start_month', 'start_hour', 
                                                           'morning']])
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], 
                             tmp_scaled[:idx_split,:]]))
X_test = csr_matrix(hstack([full_sites_sparse[idx_split:,:], 
                            tmp_scaled[idx_split:,:]]))

# Train the model on the whole training data set using optimal regularization parameter
lr = LogisticRegression(C=C, random_state=17, solver='liblinear').fit(X_train, y_train)

# Make a prediction for the test set
y_test = lr.predict_proba(X_test)[:, 1]

# Write it to the submission file
write_to_submission_file(y_test, 'baseline_2.csv')

In [ ]:
################ ######## ######## ######## ######## ######## ########  baseline

# Part 2. Freeride

<img src='../../img/snowboard.jpg' width=70%>

*Yorko in Sheregesh, the best place in Russia for snowboarding and skiing.*

In this part, you'll need to beat the 2 more baselines mentioned in the beginning of this assignment. No more step-by-step instructions. But it'll be very helpful for you to study the Kernel "[Correct time-aware cross-validation scheme](https://www.kaggle.com/kashnitsky/correct-time-aware-cross-validation-scheme)".

Here are a few tips for finding new features: think about what you can come up with using existing features, try multiplying or dividing two of them, justify or decline your hypotheses with plots, extract useful information from time series data (time1 ... time10), do not hesitate to convert an existing feature (for example, take a logarithm), etc. Checkout other [Kernels](https://www.kaggle.com/c/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking2/kernels). We encourage you to try new ideas and models - it's fun!

*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a2_alice__*

When you get into Kaggle and Xgboost, you'll feel like that, and it's OK :)

<img src='../../img/xgboost_meme.jpg' width=50%>

*More nerdy jokes in the thread ["Nerd Laughing Loud"](https://www.kaggle.com/general/76963)*

# Relax? Not yet

That's it for this part of assignment 2. More credits will be given to the top 10 performers in this competition (from among current participants), check [course roadmap](https://mlcourse.ai/roadmap). Do not spoil the assignment and the competition - don't share high-performing kernels (with LB ROC AUC >= 0.95).